### PR TITLE
Add known issues to Summary page and details for DateTime formatting

### PR DESCRIPTION
**Pull request by sven-oly**
This now shows two types of known issues, removing them from test failures and showing test results in orange.

There isn't yet a URL or other explanation of the known issue being reported, however.

### CODE REVIEW

**Comment by sven-oly** [comment]
Ready for you review. Known issues now can be shown in the UI and are characterized.

Only two are identified now:
1. NBSP replaced by ASCII space in date/time 
2. non-ASCII digits replaced by ASCII

There's not yet a way to show details or have URL pointing to more information.

File: executors/rust/1.3/src/relativedatetimefmt.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
+ // https://docs.rs/icu/1.3.2/icu/datetime/struct.DateTimeFormatter.html
```
Comment (sffc) [inline executors/rust/1.3/src/relativedatetimefmt.rs:1]: (unrelated file)

File: known_issues/known_issue_data.csv [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# File of known issues in ICU/conformance matrix
```
Comment (sffc) [inline known_issues/known_issue_data.csv:1]: Nit: Don't add files that we aren't using because it is confusing; people will find this file and think it does something but it doesn't

**Review by sffc — COMMENTED** [review]
.

File: executors/rust/1.3/src/relativedatetimefmt.rs (hunk 1 above)
Comment (sven-oly) [reply executors/rust/1.3/src/relativedatetimefmt.rs]: Removed

File: known_issues/known_issue_data.csv (hunk 2 above)
Comment (sven-oly) [reply known_issues/known_issue_data.csv]: Removed for now.

File: verifier/check_known_issues.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        # where other digits were expected, return True
+        sm = SequenceMatcher(None, expected, actual)
+        sm_opcodes = sm.get_opcodes()
+
+        digit_replace = False
+        non_digit_replacement = False
+        for diff in sm_opcodes:
+            kind = diff[0]
+            old_val = expected[diff[1]:diff[2]]
+            new_val = actual[diff[1]:diff[2]]
```
Comment (echeran) [inline verifier/check_known_issues.py:80]: please fix indices according to how the 4 indices contained within the 5-tuples returned by `get_opcodes` correspond to the 2 input strings that you originally provide as input:
https://docs.python.org/3/library/difflib.html#difflib.SequenceMatcher.get_opcodes

File: verifier/check_known_issues.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                if old_val.isdigit() and new_val.isdigit():
+                    # TODO!! : check the value of the numeral
+                    # If the same value, then its a numbering system difference
+                    if unicodedata.numeric(old_val) == unicodedata.numeric(new_val):
+                        digit_replace = True
+                    else:
+                        # Both were digits but different numeric values
+                        non_digit_replacement = True
+                else:
+                    # a digit was replaced with a non-digit
```
Comment (echeran) [inline verifier/check_known_issues.py:91]: the comment assumes too much. the difference could also be from 2 substrings totally unrelated to digits.
```suggestion
                    # the difference is not caused just by differing numbering systems
```

File: verifier/check_known_issues.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+import unicodedata
+
+# Handle known issues database
+
+# Automatically compute some known issues with certain patterns of differences
+# in actual output vs. expected
+
+# E.g., NBSP vs SP in NodeJS DateTime in ICU73, ...
+
+class known_issue():
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply verifier/check_known_issues.py:31]: Yup, will do!

File: verifier/check_known_issues.py (hunk 4 above)
Comment (sven-oly) [reply verifier/check_known_issues.py:91]: OK

File: verifier/check_known_issues.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+SP = '\u0020'
+
+# Global KI Info.
+# TODO: add this to a class
+known_issue_nbsp_sp = 'ASCII Space instead of NBSP'
+known_issue_replaced_numerals = 'Not creating non-ASCII numerals'
+
+# TODO! Load known issues from file of known problems rather than hardcoding
+
+class compute_known_issue():
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply verifier/check_known_issues.py:55]: PTAL

File: verifier/check_known_issues.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sven-oly) [reply verifier/check_known_issues.py:31]: Done

File: verifier/check_known_issues.py (hunk 3 above)
Comment (sven-oly) [reply verifier/check_known_issues.py:80]: Added text to clarify.

File: verifier/check_known_issues.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                        # Both were digits but different numeric values
+                        non_digit_replacement = True
+                else:
+                    # a digit was replaced with a non-digit
+                    non_digit_replacement = True
+
+        # Only true if the only changes were replacing digits
+        return digit_replace and not non_digit_replace
+
+def check_datetime(test_results):
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply verifier/check_known_issues.py:97]: PTAL

**Comment by sven-oly** [comment]
Updated today. PTAL!

File: verifier/check_known_issues.py [hunk 8]
```diff
@@ -28,125 +33,142 @@
...
+class knownIssueType(Enum):
+    known_issue_nbsp_sp = 'ASCII Space instead of NBSP'
+    known_issue_replaced_numerals = 'Not creating non-ASCII numerals'
 
 # TODO! Load known issues from file of known problems rather than hardcoding
 
-class compute_known_issue():
-    def __init__(self):
-        return
+def compute_known_issues(test_type, test, known_issue_list):
```
Comment (echeran) [inline verifier/check_known_issues.py:48]: remove `known_issue_list` from parameters here. instead, append to `known_issues_list` in the caller (`compute_known_issues`)

File: verifier/check_known_issues.py [hunk 9]
```diff
@@ -28,125 +33,142 @@
...
+    # Look at the array of test result types, failure, error, unsupported
+    # Extract any tests from these that are known issues
+    # Return the list of tests that are known issues
+    #
     known_issues_list = []
-    comp_ki = compute_known_issue()
-    for category in test_results:
+
+    for category in test_results_to_check:
         indices_to_remove = set()
```
Comment (echeran) [inline verifier/check_known_issues.py:159]:
```suggestion
        test_indices_with_known_issues = set()
```

File: verifier/check_known_issues.py [hunk 10]
```diff
@@ -28,125 +33,142 @@
...
+    if remove_this_one:
+        known_issues_list.append(test_data)
+
+    return remove_this_one
 
-        # Only true if the only changes were replacing digits
-        return digit_replace and not non_digit_replace
 
-def check_datetime(test_results):
+def check_issues(test_type, test_results_to_check):
```
Comment (echeran) [inline verifier/check_known_issues.py:151]: reminder: maybe organize code from high-level at the bottom of the file to low-level code at the top of the file

File: verifier/check_known_issues.py (hunk 9 above)
Comment (sven-oly) [reply verifier/check_known_issues.py:159]: Renamed

File: verifier/check_known_issues.py [hunk 11]
```diff
@@ -119,35 +107,44 @@ def numerals_replaced(expected, actual):
         return None
 
 
-def check_datetime_test(test_data, known_issues_list):
+def check_datetime_known_issues(test_data):
```
Comment (echeran) [inline verifier/check_known_issues.py:110]: rename all occurrences of `test_data` to `test` in order to be consistent with the other functions & their local variables.

File: verifier/check_known_issues.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    # See https://docs.python.org/3/library/difflib.html#difflib.SequenceMatcher.get_opcodes
+    sm = SequenceMatcher(None, expected, actual)
+    sm_opcodes = sm.get_opcodes()
+
+    digit_replace = False
+    non_digit_replacement = False
+
+    # sm_opcodes describe the changes to turn expected string into the actual string
+    # See https://docs.python.org/3/library/difflib.html#difflib.SequenceMatcher.get_opcodes
+    # The tuple is [tag, i1, i2, j1, k2]
```
Comment (echeran) [inline verifier/check_known_issues.py:80]:
```suggestion
    # The tuple is [tag, i1, i2, j1, j2]
```

**Review by echeran — APPROVED** [review]
LGTM
